docs/source/settings.rst
App-specific Settings
=====================

**django-auth_enhanced** provides various settings to be as flexible and
pluggable as possible.

This section lists all available settings with their respective default values.
To alter any of these settings, simply include them in your project's settings
module.

**django-auth_enhanced** will check its settings automatically for validity
and report any errors with detailled hints.


Available Settings
------------------

.. glossary::

    DAE_ADMIN_LIST_DISPLAY
        Controls the values, that are included in Django's admin backend object
        list view. Django's default ``UserAdmin`` implementation uses the
        following list, which is considered the *default value*: ``('username', 'email', 'first_name', 'last_name', 'is_staff')``.

        If this setting value is set, it will simply substitute the ``list_display``-attribute
        of the admin class.

        If your project uses a custom user model, make sure to include actual
        attributes of that model in this list.

        Additionally, the following enhancements are provided by the app's
        implementation of the admin class:

        * ``'is_active_with_action'``: Combines the value of ``is_active`` with a button to activate / deactive this user.
        * ``'status_aggregated'``: Returns localized strings, describing the status of the user. Possible values: ``'user'``, ``'staff'`` and ``'superuser'``.
        * ``'username_status_char'``: Returns the username attribute, prefixed with a configurable character. See :term:`DAE_ADMIN_USERNAME_STATUS_CHAR`.
        * ``'username_status_color'``: Returns the username, colored with configurable colors. See :term:`DAE_ADMIN_USERNAME_STATUS_COLOR`.
        * ``'toggle_is_active'``: Shows a button to activate / deactivate the user, depending on his activation status (``is_active``). It is recommended to use ``'is_active_with_action'`` instead.

        **Accepted Values:**

        * a tuple containing attributes of the user model as strings, e.g. ``('username', 'email')``.

    DAE_ADMIN_SEARCH_FIELDS
        Controls the fields, that are considered when using the search box in
        Django's admin backend object list view. Django's default ``UserAdmin``
        implementation uses the following list, which is considered the
        *default value*: ``('username', 'email', 'first_name', 'last_name')``.

        If this setting value is set, it will simply substitute the ``search_fields``-attribute
        of the admin class.

        If your project uses a custom user model, make sure to include actual
        attributes of that model in this list.

        Furthermore, if set to an empty tuple, ``()``, the search box will be
        disabled.

        **Accepted Values:**

        * a tuple containing attributes of the user model as strings, e.g. ``('username', 'email')``.
        * an empty tuple ``()``. This deactivates the search box.

    DAE_ADMIN_SIGNUP_NOTIFICATION
        Controls, if admins / superusers will be notified of new signups.

        **Accepted Values:**

        * ``False`` (default value): No notification will be sent.
        * a tuple of the following structure: ``('django', 'django@localhost', ('mail', )),``, where ``'django'`` is a username, ``'django@localhost'`` a valid email address and ``('mail', )`` a tuple of notification methods. As of now, only ``'mail'`` is supported.

    DAE_ADMIN_USERNAME_STATUS_CHAR
        This setting controls the characters, that are used to prefix usernames
        depending on their status, if ``'username_status_char'`` is in
        :term:`DAE_ADMIN_LIST_DISPLAY`.

        **Accepted Value:**

        * a tuple containing two single characters, e.g. ``('#', '$')``. ``'#'`` will be used to prefix superusers and ``'$'`` for staff users.

    DAE_ADMIN_USERNAME_STATUS_COLOR
        This setting controls the color, that is used to color usernames,
        depending on their status, if ``'username_status_color'`` is in
        :term:`DAE_ADMIN_LIST_DISPLAY`.

        **Accepted Value:**

        * a tuple containing two hexadecimal color codes (with leading ``#``), e.g. ``('#cc0000', '#00cc00')``. Names of superusers will be colored red (``'#cc0000'``), staff users green (``'#00cc00'``).

    DAE_EMAIL_ADMIN_NOTIFICATION_PREFIX
        All emails to admins / superusers will have a subject, that is prefixed
        with this string, put in ``[]``, i.e. if this setting is set to ``'foo'``,
        the subject of mails will be ``[foo] Some generic mail to admins``. If
        this setting is set to ``''``, the subject will be ``Some generic mail
        to admins``.

        **Accepted Values:**

        * a string (default value ``''``)

    DAE_EMAIL_FROM_ADDRESS
        All emails sent by this app will use this *from*-address.

        Please note, that the default value relies on Django's
        ``DEFAULT_FROM_EMAIL``, see `Django's documentation <https://docs.djangoproject.com/en/dev/ref/settings/#default-from-email>`_
        for more details.

        **Accepted Values:**

        * all valid mail addresses

    DAE_EMAIL_PREFIX
        All emails sent to *normal users* will have a subject, that is prefixed
        with this string, put in ``[]``, i.e. if this setting is set to ``'foo'``,
        the subject of mails will be ``[foo] Some generic mail to users``. If
        this setting is set to ``''``, the subject will be ``Some generic mail
        to users``.

        **Accepted Values:**

        * a string (default value ``''``)

    DAE_EMAIL_TEMPLATE_PREFIX
        This setting determines, where to look for email templates.

        **django-auth_enhanced** makes use of Django's template engine to create
        its email messages. This setting allows you, to seperate your email
        templates from the html-templates.

        **Example:** Let's assume your Django project is set up to look for
        templates in the following path ``project_root/templates/``. By putting
        this setting to ``auth_enhanced/mail``, Django will search the path
        ``project_root/templates/auth_enhanced/mail/`` for mail templates.

        **Accepted Values:**

        * a string, that can be suffixed to a path. Please note, that this **must not include** a trailing slash (``'mail'`` instead of ``'mail/'``).

    DAE_OPERATION_MODE
        This is the most important setting of **django-auth_enhanced**,
        determing how newly registered users are handled.

        **Accepted Values:**

        * ``'auto'`` (default value): This automatic mode is the closest to Django's default behaviour. Newly registered users are activated by default and are immediatly able to login.
        * ``'email-verification'``: In this mode, the user is required to verify his email address. An automatically generated email is sent, including a verification link/token. His account is activated when the address is verified. This mode will automatically include an email field in the signup form.
        * ``'manual'``: This mode requires manual activation of newly created users. Admins/superusers will have to log into the administration backend and activate the user.

    DAE_SALT
        This salt is used to seperate different signing processes in your
        project nicely seperated. See `Django's documentation <https://docs.djangoproject.com/en/dev/topics/signing/#using-the-salt-argument>`_
        for more details.

        **Accepted Values:**

        * a string (default value ``'django-auth_enhanced'``)

    DAE_VERIFICATION_TOKEN_MAX_AGE
        This setting determines, how long any verification token is considered
        valid in the application.

        **Accepted Values:**

        * an integer, specifying the maximum age in seconds
        * a string, with ``h`` as its last character and any number of digits, that is parsable into an integer value. This will be calculated as a given amount of hours, i.e. ``'2h'`` means two hours
        * a string, with ``d`` as its last character and any number of digits, that is parsable into an integer value. This will be calculated as a given amount of days, i.e. ``'1d'`` means one day or 24 hours

        The default value is ``3600``, so all tokens are valid for one hour.


Developer's Description
-----------------------

**django-auth_enhanced** *injects* the default values of its settings on
startup, using the ``AppConfig.ready()``-method (**Yes**, this is
discouraged explicitly in Django's documentation, but it is easily the best way
to provide default settings for app-specific settings).

If you want to have a look into it: the *AppConfig* ``AuthEnhancedConfig`` is
found in ``apps.py``. However, the actual injection-functions are found in
``settings.py``.

The checks for validity are performed by Django's built-in check framework.
The app-specific checks can be found in ``checks.py``.


auth_enhanced/settings.py
# -*- coding: utf-8 -*-
"""Contains all app-specific settings with their respective default value.

Please note, that this file is more or less just a glossary of the settings,
that can be used in a project's settings module.
The 'injection' of these settings (meaning: providing their default value,
if not provided through a project's settings module) is done in this app's
'AppConfig'-class (see 'apps.py')."""


# Django imports
from django.conf import settings
from django.core.exceptions import ImproperlyConfigured
from django.utils.translation import ugettext_lazy as _  # noqa

# app imports
from auth_enhanced.exceptions import AuthEnhancedConversionError

# #############################################################################
# CONSTANTS
# #############################################################################

# the default location for mail templates
#   In this app, this means 'auth_enhanced/templates/auth_enhanced/mail/'.
#   Please note: There is no trailing slash!
DAE_CONST_EMAIL_TEMPLATE_PREFIX = 'auth_enhanced/mail'

# This mode will automatically activate newly registered accounts
DAE_CONST_MODE_AUTO_ACTIVATION = 'auto'

# This mode sends a verification email and will automaticall activate the
#   newly registered account after successfull verification
DAE_CONST_MODE_EMAIL_ACTIVATION = 'email-verification'

# This mode will *NOT* activate newly registered accounts and relies on manual
#   activation by a superuser
DAE_CONST_MODE_MANUAL_ACTIVATION = 'manual'

# contains the app's operation modes to be used in a model (choices) field
# DAE_CONST_OPERATION_MODE_CHOICES = (
#     (DAE_CONST_MODE_AUTO_ACTIVATION, _('Automatic activation')),
#     (DAE_CONST_MODE_MANUAL_ACTIVATION, _('Manual activation')),
#     (DAE_CONST_MODE_EMAIL_ACTIVATION, _('EMail activation')),
# )

# the name of the login url, as specified in 'urls.py'
DAE_CONST_RECOMMENDED_LOGIN_URL = 'auth_enhanced:login'

# this is the default value for DAE_VERIFICATION_TOKEN_MAX_AGE. It is directly
#   given in seconds, because it is the fallback value used in AppConfig
DAE_CONST_VERIFICATION_TOKEN_MAX_AGE = 3600


# #############################################################################
# FUNCTIONS
# #############################################################################

def convert_to_seconds(time_str):
    """Converts a string to a time-integer, specifying the seconds."""

    try:
        if time_str[-1:] == 'h':
            seconds = int(time_str[:-1]) * 3600
        elif time_str[-1:] == 'd':
            seconds = int(time_str[:-1]) * 3600 * 24
        else:
            seconds = int(time_str)
    except ValueError:
        raise AuthEnhancedConversionError(_("Could not convert the parameter to an integer value."))

    return seconds


def inject_setting(name, default_value):
    """Injects an app-specific setting into Django's settings module.

    If the setting is already present in the project's settings module, the set
    value will be used. If it is not present, it will be injected.

    The function will validate, that only uppercase setting-names will be used
    or raise an appropriate exception."""

    # check, that the name is uppercased
    if not name.isupper():
        raise ImproperlyConfigured('Only uppercase names are allowed!')

    # set the setting, if it is not already present
    if not hasattr(settings, name):
        setattr(settings, name, default_value)


def set_app_default_settings():
    """Sets all app-specific default settings."""

    # ### DAE_ADMIN_SHOW_SEARCHBOX
    # This setting controls, if the searchbox in admin's list view will be
    #   activated.
    # Possible values:
    #   False or True (default value, like Django default)
    #   In case of True, Django's default value for the fields to search in
    #   is used.
    inject_setting('DAE_ADMIN_SHOW_SEARCHBOX', True)

    # ### DAE_ADMIN_SIGNUP_NOTIFICATION
    # This setting controls, if emails will be sent to specified superusers/
    #   admins whenever a new user signs up.
    # Possible values:
    #   False
    #       - no emails will be sent (default value)
    #   List of tuples of the following form
    #       - (USERNAME, EMAIL_ADDRESS, (NOTIFICATION_METHOD, )),
    #       - ('django', 'django@localhost', ('mail', )),
    #           USERNAME must be a valid username of this Django project
    #           EMAIL_ADDRESS must be the verified email address of that user
    #           (NOTIFICATION_METHOD, ) must be a tuple containing supported
    #               notification methods. As of now, only 'mail' is supported
    inject_setting('DAE_ADMIN_SIGNUP_NOTIFICATION', False)

    # ### DAE_EMAIL_ADMIN_NOTIFICATION_PREFIX
    # Mails sent to superusers by django-auth_enhanced will contain a subject
    #   with this customizable prefix.
    inject_setting('DAE_EMAIL_ADMIN_NOTIFICATION_PREFIX', '')

    # ### DAE_EMAIL_FROM_ADDRESS
    # Mails sent by django-auth_enhanced will have the following 'from'-address.
    #   The default value relies on Django's DEFAULT_FROM_EMAIL-setting, which
    #   defaults to 'webmaster@localhost' itsself.
    #   You may choose to change the built-in Django-setting or this app's one.
    inject_setting('DAE_EMAIL_FROM_ADDRESS', settings.DEFAULT_FROM_EMAIL)

    # ### DAE_EMAIL_PREFIX
    # Mails sent by django-auth_enhanced will be prefixed with this string.
    #   Please note: This does not include emails to superusers, which may use
    #   'DAE_EMAIL_ADMIN_NOTIFICATION_PREFIX'.
    inject_setting('DAE_EMAIL_PREFIX', '')

    # ### DAE_EMAIL_TEMPLATE_PREFIX
    # This setting determines the place to look for mail templates.
    # Please note, that the path must be reachable by Django's template engine.
    # Furthermore, it *must not* include a trailing slash.
    inject_setting('DAE_EMAIL_TEMPLATE_PREFIX', DAE_CONST_EMAIL_TEMPLATE_PREFIX)

    # ### DAE_OPERATION_MODE
    # This setting determines the way newly registered are handled.
    # Possible values:
    #   DAE_CONST_MODE_AUTO_ACTIVATION (default value)
    #       - newly created accounts will automatically get activated.
    #   DAE_CONST_MODE_EMAIL_ACTIVATION
    #       - newly created accounts will receive a verification mail. The
    #           account get activated automatically, when the email address
    #           is verified
    #   DAE_CONST_MODE_MANUAL_ACTIVATION
    #       - this mode will *NOT* activate newly registered accounts and
    #           relies on manual activation by a superuser
    inject_setting('DAE_OPERATION_MODE', DAE_CONST_MODE_AUTO_ACTIVATION)

    # ### DAE_SALT
    # This salt is used to keep signing processes thoughout your project
    #   nicely seperated. See https://docs.djangoproject.com/en/dev/topics/signing/#using-the-salt-argument
    inject_setting('DAE_SALT', 'django-auth_enhanced')

    # ### DAE_VERIFICATION_TOKEN_MAX_AGE
    # This setting determines, how long any verification token is considered
    #   valid.
    # Possible values:
    #   - an integer, specifying the maximum age of the token in seconds
    inject_setting('DAE_VERIFICATION_TOKEN_MAX_AGE', DAE_CONST_VERIFICATION_TOKEN_MAX_AGE)
